undertow/src/main/java/org/wildfly/extension/undertow/filters/FilterCapabilities.java

```diff
@@ -0,0 +1,90 @@
+ * distribution for a full listing of individual contributors.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package org.wildfly.extension.undertow.filters;
+
+import static org.wildfly.extension.undertow.Capabilities.CAPABILITY_FILTER;
+import static org.wildfly.extension.undertow.Capabilities.CAPABILITY_FILTER_REF;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.function.Function;
+
+import org.jboss.as.controller.PathAddress;
+import org.jboss.as.controller.PathElement;
+import org.jboss.as.controller.capability.RuntimeCapability;
+import org.jboss.as.controller.descriptions.ModelDescriptionConstants;
+import org.wildfly.extension.undertow.UndertowFilter;
+import org.jboss.as.clustering.controller.Capability;
+import org.jboss.as.clustering.controller.UnaryCapabilityNameResolver;
+
+import io.undertow.server.HandlerWrapper;
+
+/**
+ * @author Paul Ferraro
+ * @author baranowb
+ */
+enum FilterCapabilities implements Capability {
+
+    FILTER_CAPABILITY(CAPABILITY_FILTER, HandlerWrapper.class),
+    FILTER_REF_CAPABILITY(CAPABILITY_FILTER_REF, UndertowFilter.class,new Function<PathAddress, String[]>() {
+
+        @Override
+        public String[] apply(PathAddress t) {
+            // /subsystem=undertow/server=default-server/host=fridge-server/filter-ref=XXX
+            // org.wildfly.extension.undertow.filter-ref.default-server.fridge-server.XXXX
+            // /subsystem=undertow/server=default-server/host=fridge-server/location=test/filter-ref=XXX
+            // org.wildfly.extension.undertow.filter-ref.default-server.fridge-server.test.XXXX
+
+            final List<String> elements = new ArrayList<String>();
+            boolean include = false;
+            final Iterator<PathElement> it = t.iterator();
+            while(it.hasNext()) {
+                PathElement e = it.next();
+                if(include) {
+                    elements.add(e.getValue());
+                } else if(e.getKey().equals(ModelDescriptionConstants.SERVER)){
+                    elements.add(e.getValue());
+                    include = true;
+                }
+            }
+            return elements.toArray(new String[elements.size()]);
+        }
+    });
```

Comment:
I do not think you can reuse the same RuntimeCapability implementation for both contexts (i.e. per-host vs per-location), since the name resolution logic if different, as is the number of dynamic segments.  This will otherwise break galleon.  I ran into a similar problem when intitially modeling capabilities for default caches in the Infinispan subsystem.  Essentially, a capability with a given name must have a fixed number of name segments.

Rather, you'll want to parameterize the FilterRefDefinition with its RuntimeCapability.  That way the host can register the variant with 3 dynamic name segments, and the location can register the variant with 4 dynamic name segments.

FYI, once https://github.com/wildfly/wildfly-core/pull/5516 is merged, you can use [1] for the host filter-ref and [2] for the location filter-ref.

[1] https://github.com/pferraro/wildfly-core/blob/WFCORE-6311/controller/src/main/java/org/jboss/as/controller/capability/TernaryCapabilityNameResolver.java#L27
[2] https://github.com/pferraro/wildfly-core/blob/WFCORE-6311/controller/src/main/java/org/jboss/as/controller/capability/QuaternaryCapabilityNameResolver.java#L28

Reply:
Ok, I will rebase it once #5516 is in.